### PR TITLE
Index column sort order #1661

### PR DESCRIPTION
**Pull request by JRDuncan**
Index column sort order

### CODE REVIEW

**Comment by othercorey** [comment]
Need to revert all of the extra formatting changes and docker compose changes.

**Comment by othercorey** [comment]
Please review the diff and remove all formatting changes.

**Comment by JRDuncan** [comment]
> remove all formatting changes

I remove changes. I'm now getting @stickler-ci
stickler-ci Failing after 3s — Review failed

`        return sprintf(
            'CREATE %s INDEX %s ON %s (%s);',
            ($index->getType() === Index::UNIQUE ? 'UNIQUE' : ''),
            $indexName,
            $this->quoteTableName($tableName),
            implode(',', $columnNames)
        );
`

**Comment by MasterOdin** [comment]
Merging master into your branch should I think fix stickler-ci on failing on that line.

File: src/Phinx/Db/Adapter/SqlServerAdapter.php [hunk 1]
```diff
@@ -1262,24 +1262,30 @@ protected function getColumnSqlDefinition(Column $column, $create = true)
...
-            $columnNames = $index->getColumns();
             $indexName = sprintf('%s_%s', $tableName, implode('_', $columnNames));
         }
+        $order = $index->getOrder();
+        if (!empty($order)) {
+            foreach ($order as $key => $value) {
+                    $loc = array_search($key, $columnNames);
+                    $columnNames[$loc] = sprintf('[%s] %s', $key, $value);
+            }
+        }
```
Comment (MasterOdin) [inline src/Phinx/Db/Adapter/SqlServerAdapter.php:1281]:
```suggestion
        $order = $index->getOrder() ?? [];
        $columnNames = array_map(function($columnName) {
            $ret = '[' . $columnName . ']';
            if (isset($order[$columnName])) {
                $ret .= ' ' . $order[$columnName];
            }
            return $ret;
        }, $columnNames);
```

This I think keeps your desired behavior, while retaining existing behavior when not using an order with regards to placing `[` and `]` around each column name.

Comment (MasterOdin) [reply src/Phinx/Db/Adapter/SqlServerAdapter.php]: @JRDuncan ack, sorry, I realize now I forgot to include `use ($order)` in the anonymous function definition when I wrote this out.

Comment (JRDuncan) [reply src/Phinx/Db/Adapter/SqlServerAdapter.php]: Thanks @MasterOdin  When I pulled It I realized it wasn't working.

**Comment by JRDuncan** [comment]
@MasterOdin @dereuromark @othercorey 
Test fails on Mysql versions less than 8 as feature was not implemented before Mysql 8. What should I do?

File: src/Phinx/Db/Adapter/MysqlAdapter.php [hunk 2]
```diff
@@ -1351,18 +1351,30 @@ protected function getIndexSqlDefinition(Index $index)
...
+        $columnNames = $index->getColumns();
+        $order = $index->getOrder() ?? [];
+        $columnNames = array_map(function ($columnName) use ($order) {
+            $ret = '`' . $columnName . '`';
+            if (isset($order[$columnName])) {
+                $ret .= ' ' . $order[$columnName];
+            }
+
+            return $ret;
+        }, $columnNames);
```
Comment (MasterOdin) [inline src/Phinx/Db/Adapter/MysqlAdapter.php:1363]: Similar to how SRID is handled for the column definition for geometry types, we can gate keep whether or not to add the order clause through a `version_compare`.

```suggestion
        $orderSupported = version_compare($this->getAttribute(\PDO::ATTR_SERVER_VERSION), '8') > -1;
        $columnNames = array_map(function ($columnName) use ($order) {
            $ret = '`' . $columnName . '`';
            if ($orderSupported && isset($order[$columnName])) {
                $ret .= ' ' . $order[$columnName];
            }

            return $ret;
        }, $columnNames);
```

File: tests/Phinx/Db/Adapter/MysqlAdapterTest.php [hunk 3]
```diff
@@ -1231,6 +1231,25 @@ public function testAddIndex()
         $this->assertTrue($table->hasIndex('email'));
     }
 
+    public function testAddIndexWithSort()
+    {
```
Comment (MasterOdin) [inline tests/Phinx/Db/Adapter/MysqlAdapterTest.php:1235]:
```suggestion
    {
        $this->adapter->connect();
        if (!$this->usingMysql8()) {
            $this->markTestSkipped('Cannot test index order on mysql versions less than 8');
        }
```

**Review by MasterOdin — CHANGES_REQUESTED** [review]

File: src/Phinx/Db/Adapter/MysqlAdapter.php (hunk 2 above)
Comment (JRDuncan) [reply src/Phinx/Db/Adapter/MysqlAdapter.php:1363]: Thanks @MasterOdin

**Review by JRDuncan — COMMENTED** [review]
All necessary changes have been made. Ready for final review

**Review by JRDuncan — COMMENTED** [review]
Review complete

**Review by MasterOdin — APPROVED** [review]

File: tests/Phinx/Db/Adapter/PostgresAdapterTest.php [hunk 4]
```diff
@@ -979,6 +979,48 @@ public function testAddIndex()
...
+                        CROSS JOIN LATERAL unnest (i.indkey) WITH ORDINALITY AS c (colnum, ordinality)
+                        LEFT JOIN LATERAL unnest (i.indoption) WITH ORDINALITY AS o (option, ordinality)
+                        ON c.ordinality = o.ordinality
+                        JOIN pg_attribute AS a ON trel.oid = a.attrelid AND a.attnum = c.colnum
+                        WHERE trel.relname = 'table1'
+                        AND irel.relname = 'table1_email_username'
+                        AND a.attname = 'username'
+                        GROUP BY o.option, tnsp.nspname, trel.relname, irel.relname");
+        $emailOrder = $rows[0];
+        $this->assertEquals($emailOrder['sort_order'], 'ASC');
```
Comment (dereuromark) [inline tests/Phinx/Db/Adapter/PostgresAdapterTest.php:1021]: We will have to fix the order here, also use assertSame().

**Pull request merged**